File: app/src/main/java/com/neeplayer/AlbumAdapter.java
package com.neeplayer;

import android.content.Context;
import android.graphics.BitmapFactory;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.nostra13.universalimageloader.core.DisplayImageOptions;
import com.nostra13.universalimageloader.core.ImageLoader;
import com.nostra13.universalimageloader.core.ImageLoaderConfiguration;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;

public class AlbumAdapter extends BaseAdapter {
    private ArrayList<Album> albums;
    private LayoutInflater inflater;
    private ImageLoader imageLoader;

    public AlbumAdapter(Context context, ArrayList<Album> albums) {
        this.albums = albums;
        inflater = LayoutInflater.from(context);

        imageLoader = ImageLoader.getInstance();
    }

    @Override
    public int getCount() {
        return albums.size();
    }

    @Override
    public Object getItem(int position) {
        return null;
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    static class ViewHolder {
        TextView title;
        TextView year;
        TextView info;
        ImageView art;
        LinearLayout songs;
    }

    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        ViewHolder holder;

        if (convertView == null) {
            convertView = inflater.inflate(R.layout.album, parent, false);
            holder = new ViewHolder();

            holder.title = (TextView) convertView.findViewById(R.id.album_title);
            holder.year  = (TextView) convertView.findViewById(R.id.album_year);
            holder.info  = (TextView) convertView.findViewById(R.id.album_info);
            holder.art  = (ImageView) convertView.findViewById(R.id.album_art);
            holder.songs = (LinearLayout) convertView.findViewById(R.id.song_list);

            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

        Album album = albums.get(position);

        holder.title.setText(album.getTitle());
        holder.year.setText(Integer.toString(album.getYear()));


        imageLoader.displayImage("file://" + album.getArt(), holder.art);

        long albumDuration = 0;
        holder.songs.removeAllViews();
        for (int i = 0; i < album.getSongs().size(); ++i) {
            Song song = album.getSongs().get(i);

            LinearLayout songView = (LinearLayout)inflater.inflate(R.layout.song, holder.songs, false);

            TextView track =    (TextView) songView.findViewById(R.id.song_track);
            TextView title =    (TextView) songView.findViewById(R.id.song_title);
            TextView duration = (TextView) songView.findViewById(R.id.song_duration);

            int songTrack = song.getTrack();
            if (songTrack >= 1000) {
                songTrack = songTrack % 1000;
            }

            track.setText(Integer.toString(songTrack));
            title.setText(song.getTitle());

            long ms = song.getDuration();
            albumDuration += ms;
            long min = TimeUnit.MILLISECONDS.toMinutes(ms);
            long sec = TimeUnit.MILLISECONDS.toSeconds(ms) - TimeUnit.MINUTES.toSeconds(min);
            duration.setText(String.format("%d:%02d", min, sec));

            songView.setTag(R.id.ALBUM_POSITION, position);
            songView.setTag(R.id.SONG_POSITION, i);

            holder.songs.addView(songView);
        }

        holder.info.setText(String.format(
                "%d songs, %d min",
                album.getSongs().size(),
                TimeUnit.MILLISECONDS.toMinutes(albumDuration)));

        return convertView;
    }
}

File: app/src/main/java/com/neeplayer/AlbumAdapter.kt
package com.neeplayer

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.bumptech.glide.Glide


import java.util.concurrent.TimeUnit

import kotlinx.android.synthetic.album.view.*;


class AlbumAdapter(private val context: Context, private val albums: List<Album>) : BaseAdapter() {
    private val inflater = LayoutInflater.from(context)


    override fun getCount(): Int = albums.size

    override fun getItem(position: Int): Any? = null

    override fun getItemId(position: Int): Long = 0

    internal data class ViewHolder(
            val title: TextView,
            val year: TextView,
            val info: TextView,
            val art: ImageView,
            val songs: LinearLayout
    )

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {

        val view = convertView ?: let {
            val view = inflater.inflate(R.layout.album, parent, false)
            view.tag = ViewHolder(view.album_title, view.album_year, view.album_info, view.album_art, view.song_list)
            view
        }

        val holder = view.tag as ViewHolder

        val album = albums[position]

        holder.title.text = album.title
        holder.year.text = album.year.toString()

        Glide.with(context).load("file://" + album.art).dontAnimate().into(holder.art)

        var albumDuration: Long = 0
        holder.songs.removeAllViews()

        album.songs.forEachIndexed { i, song ->
            val songView = inflater.inflate(R.layout.song, holder.songs, false) as LinearLayout

            val track = songView.findViewById(R.id.song_track) as TextView
            val title = songView.findViewById(R.id.song_title) as TextView
            val duration = songView.findViewById(R.id.song_duration) as TextView


            track.text = song.track?.mod(1000).toString()

            title.text = song.title

            val ms = song.duration
            albumDuration += ms
            val min = TimeUnit.MILLISECONDS.toMinutes(ms)
            val sec = TimeUnit.MILLISECONDS.toSeconds(ms) - TimeUnit.MINUTES.toSeconds(min)
            duration.text = "%d:%02d".format(min, sec)

            songView.setTag(R.id.ALBUM_POSITION, position)
            songView.setTag(R.id.SONG_POSITION, i)

            holder.songs.addView(songView)
        }

        holder.info.text = "%d songs, %d min".format(album.songs.size, TimeUnit.MILLISECONDS.toMinutes(albumDuration))

        return view
    }
}
